File: dsa/binary-search/binary-search.go
package main

import "fmt"

//Binary search function
func binarySearch(arr *[]int, n int) int{
	var l, h, mid = 0, 0, 0
	length := len(*arr)
	h = length - 1
	for ; l <= h ; {
		mid = (l + h) / 2
		if (*arr)[mid] == n {
			//Returns array index if target element is in the middle of the array
			return mid
		} else {
			if n < (*arr)[mid] {
				//If target element is less than middle element
				//searches left half of the array

				h = mid - 1
			} else {
				//If target element is greater than middle element
				//searches right half of the array 
				l = mid + 1
			}
		}
	}

	//Returns -1 if the target element is not found
	return -1
}

func main() {
	arr := []int{10, 20, 30, 40, 50, 60, 70, 80, 90, 100}
	n := 30
	fmt.Println(binarySearch(&arr, n))
}

File: dsa/binary-search/binarySearch.scala
object binarySearch {
    def search(arr: Arrays[Int], target: Int): Int = {
        var left = 0
        var right = arr.length - 1

        while(left<=right) {
            val mid = left + (right - left) / 2
            if(arr(mid) == target) return mid
            else if(arr(mid) < target) left = mid + 1
            else right = mid - 1
        }
        -1
    }
    def main(args: Array[String]): Unit = {
        var arr = Arrays(10, 20, 30, 40, 50, 60, 70, 80, 90, 100)
        var target = 5
        var result = search(arr, target)
        if(result == -1) println("Element not found")
        else println("Element found at index " + result)
    }
}